### ISSUE TITLE
卷积，bn，relu之后出现inf

### ISSUE THREAD

**Issue by fxwfzsxyq** [PaddlePaddle/Paddle#24173]
用动态图参照torchvision model修改了一个vgg，训练发现卷积，bn，relu之后的feature map中出现inf，不知道在那个位置计算出现了inf，导致loss为nan

**Comment by fxwfzsxyq** [comment]
import paddle
import paddle.fluid as fluid
from paddle.fluid.dygraph.nn import Conv2D, Pool2D, BatchNorm, Linear
import pdb 

class VGG(fluid.dygraph.Layer):
    def __init__(self, features, num_classes=1000):
        super(VGG, self).__init__()
        self.features = features
        self.avgpool = Pool2D(pool_size=7, pool_stride=1, pool_type='avg')
        import math
        stdv = 1.0 / math.sqrt(2048 * 1.0)
        self.classifier =  Linear(512,
                   num_classes,
                   act='softmax',
                   param_attr=fluid.param_attr.ParamAttr(
                   initializer=fluid.initializer.Uniform(-stdv, stdv)))

    def forward(self, x):
        pdb.set_trace()
        x = self.features(x)
        x = self.avgpool(x)
        x = fluid.layers.reshape(x, shape=[x.shape[0], -1])
        x = self.classifier(x)
        return x

cfgs = {
    'A': [64, 'M', 128, 'M', 256, 256, 'M', 512, 512, 'M', 512, 512, 'M'],
    'B': [64, 64, 'M', 128, 128, 'M', 256, 256, 'M', 512, 512, 'M', 512, 512, 'M'],
    'D': [64, 64, 'M', 128, 128, 'M', 256, 256, 256, 'M', 512, 512, 512, 'M', 512, 512, 512, 'M'],
    'E': [64, 64, 'M', 128, 128, 'M', 256, 256, 256, 256, 'M', 512, 512, 512, 512, 'M', 512, 512, 512, 512, 'M'],
}

def make_layers(cfg, batch_norm=False):
    layers = []
    in_channels = 3
    for v in cfg:
        if v == 'M':
            layers += [Pool2D(pool_size=2, pool_stride=2)]
        else:
            if batch_norm:
                layers += [Conv2D(in_channels, v, filter_size=3, padding=1), BatchNorm(v,act='relu')]
            else:
                layers += [Conv2D(in_channels, v, filter_size=3, padding=1,act='relu')]
            in_channels = v
    return fluid.dygraph.Sequential(*layers)

def _vgg(cfg, batch_norm, **kwargs):
    model = VGG(make_layers(cfgs[cfg], batch_norm=batch_norm), **kwargs)
    return model

def vgg11( **kwargs):
    r"""VGG 11-layer model (configuration "A") from
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg( 'A', False,**kwargs)


def vgg11_bn( **kwargs):
    r"""VGG 11-layer model (configuration "A") with batch normalization
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg('A', True, **kwargs)


def vgg13(**kwargs):
    r"""VGG 13-layer model (configuration "B")
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg( 'B', False, **kwargs)


def vgg13_bn( **kwargs):
    r"""VGG 13-layer model (configuration "B") with batch normalization
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg('B', True, **kwargs)


def vgg16( **kwargs):
    r"""VGG 16-layer model (configuration "D")
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg('D', False, **kwargs)


def vgg16_bn( **kwargs):
    r"""VGG 16-layer model (configuration "D") with batch normalization
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg('D', True, **kwargs)


def vgg19( **kwargs):
    r"""VGG 19-layer model (configuration "E")
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg( 'E', False, **kwargs)


def vgg19_bn(**kwargs):
    r"""VGG 19-layer model (configuration 'E') with batch normalization
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg( 'E', True, **kwargs)

**Comment by fxwfzsxyq** [comment]
**import paddle
import paddle.fluid as fluid
from paddle.fluid.dygraph.nn import Conv2D, Pool2D, BatchNorm, Linear
import pdb 

class VGG(fluid.dygraph.Layer):
    def __init__(self, features, num_classes=1000):
        super(VGG, self).__init__()
        self.features = features
        self.avgpool = Pool2D(pool_size=7, pool_stride=1, pool_type='avg')
        import math
        stdv = 1.0 / math.sqrt(2048 * 1.0)
        self.classifier =  Linear(512,
                   num_classes,
                   act='softmax',
                   param_attr=fluid.param_attr.ParamAttr(
                   initializer=fluid.initializer.Uniform(-stdv, stdv)))

    def forward(self, x):
        pdb.set_trace()
        x = self.features(x)
        x = self.avgpool(x)
        x = fluid.layers.reshape(x, shape=[x.shape[0], -1])
        x = self.classifier(x)
        return x

cfgs = {
    'A': [64, 'M', 128, 'M', 256, 256, 'M', 512, 512, 'M', 512, 512, 'M'],
    'B': [64, 64, 'M', 128, 128, 'M', 256, 256, 'M', 512, 512, 'M', 512, 512, 'M'],
    'D': [64, 64, 'M', 128, 128, 'M', 256, 256, 256, 'M', 512, 512, 512, 'M', 512, 512, 512, 'M'],
    'E': [64, 64, 'M', 128, 128, 'M', 256, 256, 256, 256, 'M', 512, 512, 512, 512, 'M', 512, 512, 512, 512, 'M'],
}

def make_layers(cfg, batch_norm=False):
    layers = []
    in_channels = 3
    for v in cfg:
        if v == 'M':
            layers += [Pool2D(pool_size=2, pool_stride=2)]
        else:
            if batch_norm:
                layers += [Conv2D(in_channels, v, filter_size=3, padding=1), BatchNorm(v,act='relu')]
            else:
                layers += [Conv2D(in_channels, v, filter_size=3, padding=1,act='relu')]
            in_channels = v
    return fluid.dygraph.Sequential(*layers)

def _vgg(cfg, batch_norm, **kwargs):
    model = VGG(make_layers(cfgs[cfg], batch_norm=batch_norm), **kwargs)
    return model

def vgg11( **kwargs):
    r"""VGG 11-layer model (configuration "A") from
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg( 'A', False,**kwargs)


def vgg11_bn( **kwargs):
    r"""VGG 11-layer model (configuration "A") with batch normalization
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg('A', True, **kwargs)


def vgg13(**kwargs):
    r"""VGG 13-layer model (configuration "B")
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg( 'B', False, **kwargs)


def vgg13_bn( **kwargs):
    r"""VGG 13-layer model (configuration "B") with batch normalization
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg('B', True, **kwargs)


def vgg16( **kwargs):
    r"""VGG 16-layer model (configuration "D")
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg('D', False, **kwargs)


def vgg16_bn( **kwargs):
    r"""VGG 16-layer model (configuration "D") with batch normalization
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg('D', True, **kwargs)


def vgg19( **kwargs):
    r"""VGG 19-layer model (configuration "E")
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg( 'E', False, **kwargs)


def vgg19_bn(**kwargs):
    r"""VGG 19-layer model (configuration 'E') with batch normalization
    `"Very Deep Convolutional Networks For Large-Scale Image Recognition" <https://arxiv.org/pdf/1409.1556.pdf>`_

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _vgg( 'E', True, **kwargs)**

**Comment by fxwfzsxyq** [comment]
![image](https://user-images.githubusercontent.com/18416103/80303350-709e4280-87e2-11ea-8a48-29b5decd83e0.png)

**Comment by zhiqiu** [comment]
hi，动态图的一大优点就是方便调试，你可以逐层print看看那一层出现了nan。

**Comment by BADBADBADBOY** [comment]
调试过了，backward之后，再一次forward，特征图里面就出现很大的值，到后面就inf，这是给bug吧

**Comment by BADBADBADBOY** [comment]
![image](https://user-images.githubusercontent.com/27273035/80306322-325e4e80-87f5-11ea-9646-41b0785495df.png)

**Comment by BADBADBADBOY** [comment]
![image](https://user-images.githubusercontent.com/27273035/80306327-4bff9600-87f5-11ea-9425-05e4d3b8738d.png)

**Comment by BADBADBADBOY** [comment]
感觉没毛病啊

**Comment by BADBADBADBOY** [comment]
好吧，学习率过大梯度爆炸了。。。。。。。。。。。。。

**Comment by zhiqiu** [comment]
hi，刚刚看你回复说是学习率过大的原因？

**Comment by fxwfzsxyq** [comment]
是的，学习率vgg学习率过大梯度爆炸了，我学习率设置的0.1，设置成0.001就正常跑了

**Comment by fxwfzsxyq** [comment]
和vgg没加bn也有关系，加了bn也能防止，即使0.1的学习率也可以

**Comment by zhiqiu** [comment]
> 和vgg没加bn也有关系，加了bn也能防止，即使0.1的学习率也可以

嗯嗯，如果问题解决，是否可以close issue？
